### DISCUSSION TITLE
Where are my Airflow scheduler logs?

### DISCUSSION THREAD

**Discussion by MattiaGallegati** [apache/airflow#29020] [Q&A]
Hello everyone,

I've been recently deploying a "production" Airflow instance via docker-compose.
I would like to monitor Airflow components with some external system (e.g Splunk) in order to do so I intend to collect Airflow logs from file.
I can see that the scheduler is printing usefull logs when it crashes on the stdout of the container but these logs can't be found nowhere inside the logs folder, instead I can find dozens of folder (divided per date) where i find dag_name.py.log with other logs that wich I do not understand in terms of utility.

Can anybody explain how/where can I find logs in order to monitor my scheduler instance?

Thank you

**Comment by potiuk** [comment]
Basically - wherever you direct it. Look at `airlfow scheduler --help`.

**Reply by MattiaGallegati** [reply]
Thank you,
I can see that airflow scheduler is provided with option --stdout --stderr. I've tryed redirecting logs after modifying my docker-compose command. Unfortunatelly it seems to ignore the options. Am I missing something?

For the other scheduler logs (divided per folder, named by date) is there a reference on where they come from and how to interpret them?

**Reply by potiuk** [reply]
Have you seen `--log` flag?

**Reply by potiuk** [reply]
```
usage: airflow scheduler [-h] [-D] [-p] [-l LOG_FILE] [-n NUM_RUNS]
                         [--pid [PID]] [-s] [--stderr STDERR]
                         [--stdout STDOUT] [-S SUBDIR] [-v]

Start a scheduler instance

optional arguments:
  -h, --help            show this help message and exit
  -D, --daemon          Daemonize instead of running in the foreground
  -p, --do-pickle       Attempt to pickle the DAG object to send over to the workers, instead of letting workers run their version of the code
  -l LOG_FILE, --log-file LOG_FILE
                        Location of the log file
  -n NUM_RUNS, --num-runs NUM_RUNS
                        Set the number of runs to execute before exiting
  --pid [PID]           PID file location
  -s, --skip-serve-logs
                        Don't start the serve logs process along with the workers
  --stderr STDERR       Redirect stderr to this file
  --stdout STDOUT       Redirect stdout to this file
  -S SUBDIR, --subdir SUBDIR
                        File location or directory from which to look for the dag. Defaults to '[AIRFLOW_HOME]/dags' where [AIRFLOW_HOME] is the value you set for 'AIRFLOW_HOME' config you set in 'airflow.cfg' 
  -v, --verbose         Make logging output more verbose

Signals:

  - SIGUSR2: Dump a snapshot of task state being tracked by the executor.

    Example:
        pkill -f -USR2 "airflow scheduler"
```

**Reply by potiuk** [reply]
For the other logs -> look at our documenation (https://airflow.apache.org/docs/apache-airflow/stable/logging-monitoring/index.html) - it describes logging and monitoring arrchitecture and more.

**Reply by MattiaGallegati** [reply]
Thank you so much, I saw them and finally managed to print log on files BUT as stated in #21843 it seems that these options are ignored unless you start the scheduler or the webserver in --daemon mode. Are there any particular reasons for this?

I saw  the documentation, but unfortunally the log structure of the scheduler/dag_processor_manager are not explained in detail.

**Reply by potiuk** [reply]
If you have an idea how to improve the documentation  - feel free. It is very easy - just get to the documentation page and click "suggest a change on this page" and PR will be opened and you will hace PR opened and you will be able to update the docs as easily as writing this comment (same UI). You might not realise but the code and documentation of Airlfow is created mostly by pople likel you. Airflow is created by > 2300 contributors - most of them people like you who noticed a problem or missing documentation and decided to fix it.  And likely you are the best person in the world now @MattiaGallegati to explain it in the way that other people like you will find useful and helpful. You just saw a missing piece of docs, you researched it to find the solutions and you know how it works. And you know where and how people like you would look for it. Perfect combinations. 

I will be super happy to review such documentation PRs and help you to get it merged. 

Loooking foward to it.

**Reply by potiuk** [reply]
And yes this is a standard featiure of any application that unless they have special requirements to write mutliple different log files like it is in tasks by default they output stuff to stdout and stderr. This is absolutely standard linux application behaviour. Described for example here: https://www.middlewareinventory.com/blog/linux-stdout-stderr-bash

And it is absolutely normal and expected that whoever deploys the app will redirect those outputs to files as they deem appropriate. 

This is different in case of --deamon process (described for example here) https://www.javatpoint.com/linux-daemon#:~:text=A%20daemon%20is%20also%20called,that%20manages%20the%20Apache%20server - where process is supposed to be run in unattended manner without stdout and stderr (closing those standard opened handles is part of standard "daemonization" process so there is no choice by writing the output/error to dedicated files - because stdout and stderr do not exist.

All that is a standard "linux" behaviour for any process that is supposed to be run unattended and airflow follows the same practices as any other apps and explaining all basics like that in airlfow documentation would be repeating the same information that is a standard knowledge of anyone running linux application.

But if you find it useful to be added to the docs in the form that is not repeatative, and hinting to other users like you about those - agan - you are probably the best person in the world now equipped with the knowledge you got and able to explain it to people like you to makes such documentation PR. Looking forward to it/

**Comment by MattiaGallegati** [comment]
@potiuk unforntunatelly I'm not in the condition to add strong reference to the Airflow manual, I'm still strugling to understand logging functionalities of Airflow. In some cases Airflow still behave not as I expect and I have first to understand all the reasons behind this behaviours (e.g. #29302 ).
Maybe in the future, when everything will run as I expect I can contribute.
